how-to/workspace-platform-starter/client/src/modules/actions/dynamic-menu/actions.ts

```diff
@@ -0,0 +1,135 @@
+import type OpenFin from "@openfin/core";
+import {
+	CustomActionCallerType,
+	type CustomActionPayload,
+	type CustomActionsMap,
+	type Page,
+	type WorkspacePlatformModule
+} from "@openfin/workspace-platform";
+import type { Actions } from "workspace-platform-starter/shapes/actions-shapes";
+import type { Logger, LoggerCreator } from "workspace-platform-starter/shapes/logger-shapes";
+import type { ModuleDefinition, ModuleHelpers } from "workspace-platform-starter/shapes/module-shapes";
+import { isEmpty, randomUUID } from "workspace-platform-starter/utils";
+import type { DynamicMenuProviderOptions } from "./shapes";
+
+/**
+ * Implementation for the dynamic menu actions provider.
+ */
+export class DynamicMenuProvider implements Actions<DynamicMenuProviderOptions> {
+	/**
+	 * The logger for displaying information from the module.
+	 * @internal
+	 */
+	private _logger?: Logger;
+
+	/**
+	 * Helper methods for the module.
+	 * @internal
+	 */
+	private _helpers: ModuleHelpers | undefined;
+
+	/**
+	 * Initialize the module.
+	 * @param definition The definition of the module from configuration include custom options.
+	 * @param loggerCreator For logging entries.
+	 * @param helpers Helper methods for the module to interact with the application core.
+	 * @returns Nothing.
+	 */
+	public async initialize(
+		definition: ModuleDefinition<DynamicMenuProviderOptions>,
+		loggerCreator: LoggerCreator,
+		helpers: ModuleHelpers
+	): Promise<void> {
+		this._logger = loggerCreator("DynamicMenuProvider");
+		this._helpers = helpers;
+	}
+
+	/**
+	 * Get the actions from the module.
+	 * @param platform The platform module.
+	 * @returns The map of custom actions.
+	 */
+	public async get(platform: WorkspacePlatformModule): Promise<CustomActionsMap> {
+		const actionMap: CustomActionsMap = {};
+
+		actionMap["pages-menu"] = async (payload: CustomActionPayload): Promise<void> => {
+			if (payload.callerType === CustomActionCallerType.CustomButton) {
+				const pages: Page[] = await platform.Storage.getPages();
+				const res = await this.showCustomMenu(
+					{ x: payload.x, y: payload.y },
+					payload.windowIdentity,
+					"There are no pages",
+					pages
+						.map((p) => ({
+							label: p.title,
+							id: p.pageId
+						}))
+						.sort((a, b) => a.label.localeCompare(b.label))
+				);
+
+				if (!isEmpty(res) && this._helpers?.launchPage) {
+					const pageToLaunch = await platform.Storage.getPage(res);
```

Comment:
Should this have logic borrowed from the other menu option that we added to the pages module? That has logic to show a page if it is already visible not sure that happens if you launch a page and it is already visible (maybe that logic needs pushing down into our launch page code and used by both modules).

Reply:
Isn't the behavior a bit different though, for the other module you are launching from a menu, so you know the target window identity - from a dock you don't know.

Reply:
launchPage logic has been centralized for all things launching pages Home, Menu, Dock, Share